azurelinuxagent/ga/telemetry_service.py

```diff
@@ -0,0 +1,122 @@
+# Microsoft Azure Linux Agent
+#
+# Copyright 2020 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+import datetime
+import threading
+
+import traceback
+
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.event import add_event, WALAEventOperation
+from azurelinuxagent.common.future import ustr, PriorityQueue
+from azurelinuxagent.common.interfaces import ThreadHandlerInterface
+
+
+def get_telemetry_service_handler(protocol_util):
+    return TelemetryServiceHandler(protocol_util)
+
+
+class TelemetryServiceHandler(ThreadHandlerInterface):
+    """
+    This Handler takes care of sending all telemetry out of the agent to Wireserver. It sends out data as soon as
+    there's any data available in the queue to send.
+    """
+
+    _THREAD_NAME = "TelemetryServiceHandler"
+    _MAX_TIMEOUT = datetime.timedelta(minutes=5).seconds
+
+    def __init__(self, protocol_util):
+        self._protocol = protocol_util.get_protocol()
+        self.should_run = True
+        self._thread = None
+        self._should_process_events = threading.Event()
+        self._queue = PriorityQueue()
+
+    @staticmethod
+    def get_thread_name():
+        return TelemetryServiceHandler._THREAD_NAME
+
+    def run(self):
+        logger.info("Start Extension Telemetry service.")
+        self.start()
+
+    def is_alive(self):
+        return self._thread is not None and self._thread.is_alive()
+
+    def start(self):
+        self._thread = threading.Thread(target=self._process_telemetry_thread)
+        self._thread.setDaemon(True)
+        self._thread.setName(self.get_thread_name())
+        self._thread.start()
+
+    def stop(self):
+        """
+        Stop server communication and join the thread to main thread.
+        """
+        self.should_run = False
+        if self.is_alive():
+            self.join()
+
+    def join(self):
+        self._queue.join()
+        self._thread.join()
+
+    def stopped(self):
+        return not self.should_run
+
+    def enqueue_event_func(self, event):
```

Comment:
suggestion: remove "_func" from name

Reply:
I was actually debating with myself on whether to add it or not and decided for it because I personally felt it made it obvious as to what this function pointer was doing (mainly because its passed around into different handlers). But I can remove that no issues!

Reply:
Done